lib/portlayer/event/subscriber.go

```diff
@@ -0,0 +1,208 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package event
+
+import (
+	"sync/atomic"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/vic/lib/portlayer/event/events"
+)
+
+const (
+	suspendDisabled int32 = iota
+	suspendDiscard
+	suspendQueue
+)
+
+type Subscriber interface {
+	// Topic returns the topic this subscriber is subscribed to
+	Topic() string
+	// Name returns the name of the subscriber
+	Name() string
+
+	// Suspend suspends processing events by the subscriber. If
+	// queueEvents is true, the events are queued until Resume()
+	// is called. If queueEvents is false, events passed into
+	// onEvent() after this call are discarded.
+	Suspend(queueEvents bool)
+	// Resume resumes processing of events by the subscriber.
+	// If Suspend() was called with queueEvents as true, any events
+	// that were passed to onEvent() after Suspend() returned are
+	// processed first.
+	Resume()
+	// IsSuspended returns true if the subscriber is suspended.
+	IsSuspended() bool
+
+	// Discarded returns the number of packets that were discarded by
+	// the subscriber as a result of Pause() being called with
+	// queueEvents as false.
+	Discarded() uint64
+	// Dropped returns the number of packets that were dropped when
+	// the event queue overflows. This only happens when Pause()
+	// is called with queueEvents as true.
+	Dropped() uint64
+
+	// onEvent is called by event.Manager to send an event to
+	// a subscriber
+	onEvent(events.Event)
+}
+
+type subscriber struct {
+	topic              string
+	name               string
+	callback           func(e events.Event)
+	eventQ             chan events.Event
+	suspendState       int32
+	discarded, dropped uint64
+	suspend            chan suspendCmd
+}
+
+type suspendCmd struct {
+	suspend bool
+	done    chan struct{}
+}
+
+const maxEventQueueSize = 1000
+
+// newSubscriber creates a new subscriber to topic
+func newSubscriber(topic, name string, callback func(e events.Event)) Subscriber {
+	s := &subscriber{
+		topic:    topic,
+		name:     name,
+		callback: callback,
+		eventQ:   make(chan events.Event, maxEventQueueSize),
+		suspend:  make(chan suspendCmd),
+	}
+
+	go func() {
+		suspended := false
+		var done chan struct{}
+		for {
+			if done != nil {
+				done <- struct{}{}
+				done = nil
+			}
+
+			if suspended {
+				select {
+				case c := <-s.suspend:
+					suspended = c.suspend
+					done = c.done
+				}
+
+				continue
+			}
+
+			// not suspended
+			select {
+			case e := <-s.eventQ:
+				s.callback(e)
+			case c := <-s.suspend:
+				suspended = c.suspend
+				done = c.done
+			}
```

Comment:
If suspended is false, that indicates Resume() was called before. If we end up on line 110, and Suspend() is called, line 113 should be hit, and we should set suspended to true, and loop back around.